source/io/Bootstrap.cpp

```diff
@@ -2,75 +2,88 @@
  * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
  * SPDX-License-Identifier: Apache-2.0.
  */
-#include <aws/crt/Api.h>
+#include <aws/common/clock.h>
+
 #include <aws/crt/io/Bootstrap.h>
 
+// Maximum time to wait (in seconds) for the blocking shutdown sequence.
+// This is to avoid hanging at program exit due to unexpected conditions.
+#define FINAL_TIMEOUT_SECS 10
+
 namespace Aws
 {
     namespace Crt
     {
         namespace Io
         {
-
-            /**
-             * @private
-             * Holds the bootstrap's shutdown promise.
-             * Lives until the bootstrap's shutdown-complete callback fires.
-             */
-            class ClientBootstrapCallbackData
+            namespace
             {
-              private:
-                Allocator *m_allocator;
-
-              public:
-                ClientBootstrapCallbackData(Allocator *allocator) : m_allocator(allocator) {}
-                /**
-                 * Promise for bootstrap's shutdown.
-                 */
-                std::promise<void> ShutdownPromise;
-                /**
-                 * User callback of bootstrap's shutdown-complete.
-                 */
-                OnClientBootstrapShutdownComplete ShutdownCallback;
+                /* Called by event-loop host-resolver, and client-bootstrap shutdown. */
+                void shutdown_callback(void *user_data)
+                {
+                    auto shutdown_user_data = static_cast<struct shutdown_user_data *>(user_data);
 
+                    aws_mutex_lock(&shutdown_user_data->mtx);
+                    shutdown_user_data->shutdown_count += 1;
+                    aws_condition_variable_notify_one(&shutdown_user_data->cv);
+                    aws_mutex_unlock(&shutdown_user_data->mtx);
+                }
                 /**
-                 * Internal callback of bootstrap's shutdown-complete
+                 * Return true if all three components (event_loop_group, host_resolver, client bootstrap)
+                 * have invoked the above shutdown_callback.
                  */
-                static void OnShutdownComplete(void *userData)
+                bool condition_predicate_fn(void *pred_ctx)
                 {
-                    auto callbackData = static_cast<ClientBootstrapCallbackData *>(userData);
-
-                    callbackData->ShutdownPromise.set_value();
-                    if (callbackData->ShutdownCallback)
-                    {
-                        callbackData->ShutdownCallback();
-                    }
-
-                    Crt::Delete(callbackData, callbackData->m_allocator);
+                    const auto shutdown_user_data = static_cast<struct shutdown_user_data *>(pred_ctx);
+                    return shutdown_user_data->shutdown_count == 3;
```

Comment:
Hmmm, is there an alternative strategy to counting to 3 in the shutdown callbacks ? I'm wondering if something more explicit can help, because if a developer adds another system to shut down, are they going to know to setup the shutdown callback and set this check to 4 ?

Reply:
Passing in the same callback/data into all three `aws_shutdown_callback_options` simplifies the implementation.
Setting each individual completion in a callback would require 3 different types of callbacks, each setting a different condition.
This would not solve the problem if someone modifies `aws_client_bootstrap` to include more asynchronous shutdown operations; they would still have to update the code.